stan/math/prim/fun/lb_constrain.hpp

```diff
@@ -41,21 +49,104 @@ inline auto lb_constrain(const T& x, const L& lb) {
+  }
+}
+
+/**
+ * Specialization of `lb_constrain` to apply a scalar lower bound elementwise
+ *  to each input.
+ *
+ * @tparam T A type inheriting from `EigenBase`.
+ * @tparam L Scalar.
+ * @param[in] x unconstrained input
+ * @param[in] lb lower bound on output
+ * @return lower-bound constrained value corresponding to inputs
+ */
+template <typename T, typename L, require_eigen_t<T>* = nullptr,
+          require_stan_scalar_t<L>* = nullptr,
+          require_all_not_st_var<T, L>* = nullptr>
+inline auto lb_constrain(T&& x, L&& lb) {
+  return make_holder(
+      [](const auto& x, const auto& lb) {
+        return x.unaryExpr([lb](auto&& x) { return lb_constrain(x, lb); });
+      },
+      std::forward<T>(x), std::forward<L>(lb));
+}
+
+/**
+ * Specialization of `lb_constrain` to apply a scalar lower bound elementwise
+ *  to each input.
+ *
+ * @tparam T A type inheriting from `EigenBase`.
+ * @tparam L Scalar.
+ * @param[in] x unconstrained input
+ * @param[in] lb lower bound on output
+ * @param[in,out] lp reference to log probability to increment
+ * @return lower-bound constrained value corresponding to inputs
+ */
+template <typename T, typename L, require_eigen_t<T>* = nullptr,
+          require_stan_scalar_t<L>* = nullptr,
+          require_all_not_st_var<T, L>* = nullptr>
+inline auto lb_constrain(const T& x, const L& lb,
+                         std::decay_t<return_type_t<T, L>>& lp) {
+  return eval(
+      x.unaryExpr([lb, &lp](auto&& xx) { return lb_constrain(xx, lb, lp); }));
+}
+
+/**
+ * Specialization of `lb_constrain` to apply a matrix of lower bounds
+ * elementwise to each input element.
+ *
+ * @tparam T A type inheriting from `EigenBase`.
+ * @tparam L A type inheriting from `EigenBase`.
+ * @param[in] x unconstrained input
+ * @param[in] lb lower bound on output
+ * @return lower-bound constrained value corresponding to inputs
+ */
+template <typename T, typename L, require_all_eigen_t<T, L>* = nullptr,
+          require_all_not_st_var<T, L>* = nullptr>
+inline auto lb_constrain(T&& x, L&& lb) {
+  return make_holder(
+      [](const auto& x, const auto& lb) {
+        return x.binaryExpr(
+            lb, [](auto&& x, auto&& lb) { return lb_constrain(x, lb); });
+      },
+      std::forward<T>(x), std::forward<L>(lb));
+}
+
+/**
+ * Specialization of `lb_constrain` to apply a matrix of lower bounds
+ * elementwise to each input element.
+ *
+ * @tparam T A type inheriting from `EigenBase`.
+ * @tparam L A type inheriting from `EigenBase`.
+ * @param[in] x unconstrained input
+ * @param[in] lb lower bound on output
+ * @param[in,out] lp reference to log probability to increment
+ * @return lower-bound constrained value corresponding to inputs
+ */
+template <typename T, typename L, require_all_eigen_t<T, L>* = nullptr,
+          require_all_not_st_var<T, L>* = nullptr>
+inline auto lb_constrain(const T& x, const L& lb,
+                         std::decay_t<return_type_t<T, L>>& lp) {
```

Comment:
Is the decay necessary? I feel like if it is we should probably patch `return_type_t`.

Reply:
No idt so